### PR TITLE
[AArch64][GlobalISel] Fix incorrect ABI when tail call not supported

### LINKED ISSUE #70207: [AArch64][GISel] Incorrect ABI for calls with tail marker
```llvm
; RUN: llc -O0 -mtriple=aarch64-unknown-linux-gnu < %s | FileCheck %s
declare void @func(i64, i64, i64, i64, i64, i128, i128)

define void @test(i128 %arg1, i128 %arg2) nounwind {
  tail call void @func(i64 0, i64 0, i64 0, i64 0, i64 0, i128 %arg1, i128 %arg2)
  ret void
}
```
This produces:
```asm
	sub	sp, sp, #64
	str	x30, [sp, #48]                  // 8-byte Folded Spill
	mov	x6, x0
	mov	x8, sp
	str	x1, [x8]
	str	x2, [x8, #16]
	str	x3, [x8, #32]
	mov	x4, xzr
	mov	x0, x4
	mov	x1, x4
	mov	x2, x4
	mov	x3, x4
	bl	func
	ldr	x30, [sp, #48]                  // 8-byte Folded Reload
	add	sp, sp, #64
	ret
```
Note that both of the i128 arguments get passed on the stack. Compare with the `-global-isel=0` output:
```asm
	sub	sp, sp, #32
	str	x30, [sp, #16]                  // 8-byte Folded Spill
	mov	x7, x1
	mov	x6, x0
	mov	x8, sp
	str	x3, [x8, #8]
	str	x2, [x8]
	mov	x4, xzr
	mov	x0, x4
	mov	x1, x4
	mov	x2, x4
	mov	x3, x4
	bl	func
	ldr	x30, [sp, #16]                  // 8-byte Folded Reload
	add	sp, sp, #32
	ret
```
In this case the first i128 argument is passed in registers and only the second one on the stack.

Peculiarly, this does not happen when dropping the `tail` marker (which shouldn't make a difference as no actual tail call gets generated).

### PR DESCRIPTION
**Pull request by nikic**
The check for whether a tail call is supported calls determineAssignments(), which may modify argument flags. As such, even though the check fails and a non-tail call will be emitted, it will not have a different (incorrect) ABI.

Fix this by operating on a separate copy of the arguments.

Fixes https://github.com/llvm/llvm-project/issues/70207.

### CODE REVIEW

**Comment by tschuett** [comment]
The documentation of `determineAssignments` states that it manipulates its arguments. This looks hacky. Probably `determineAssignments` should return an optional instead. Mutable arguments are an anti pattern.

**Comment by aemerson** [comment]
> The documentation of `determineAssignments` states that it manipulates its arguments. This looks hacky. Probably `determineAssignments` should return an optional instead. Mutable arguments are an anti pattern.

I don't think we have to be too dogmatic about that. Unnecessary array copies also (minutely) impact compile time which we'd be paying the cost of on every call to `determineAssignments` if we returned an optional. An alternative would be to introduce a new wrapper that's only used for querying and doesn't mutate its arguments, hiding the copying of the input from the caller. Something like `canDetermineAssignments()`?

**Comment by nikic** [comment]
Just to clarify, determineAssignments() does succeed here, it's just that following checks fail and the function has already changed state. Just a boolean canDetermineAssignments() query wouldn't be enough for use in this function, we do need the result as well.

**Comment by tru** [comment]
Should we backport this to 17.x? Remember that the next release might be the last one in the 17.x series, so we need to think carefully with what fixes we include.

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp [hunk 1]
```diff
@@ -854,7 +854,10 @@ bool AArch64CallLowering::areCalleeOutgoingArgsTailCallable(
...
   AArch64OutgoingValueAssigner CalleeAssigner(AssignFnFixed, AssignFnVarArg,
                                               Subtarget, /*IsReturn*/ false);
-  if (!determineAssignments(CalleeAssigner, OutArgs, OutInfo)) {
+  // determineAssignments() may modify argument flags, so make a copy.
+  SmallVector<ArgInfo, 8> OutArgsCopy;
+  append_range(OutArgsCopy, OutArgs);
+  if (!determineAssignments(CalleeAssigner, OutArgsCopy, OutInfo)) {
     LLVM_DEBUG(dbgs() << "... Could not analyze call operands.\n");
     return false;
   }
```
Comment (erikdesjardins) [inline llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:863]: > Just a boolean canDetermineAssignments() query wouldn't be enough for use in this function, we do need the result as well.

I interpret this to mean that the mutations performed by `determineAssignments` are necessary when this function succeeds.

But in that case, I would expect to see `*OutArgs = OutArgsCopy;` or similar at the end of this function?

**Review by aemerson — APPROVED** [review]
Let's take this change for now and leave any refactoring to another patch.

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp]: What I was referring to here is that OutInfo is used by further checks in this function. Though after your comment I saw that OutArgs is also used, so I've tweaked the code to make sure it gets the modified version, not the original one (I'm not sure whether this actually matters here, but better safe than sorry).

We don't need the result of determineAssignments() to persist outside this function (that is, the areCalleeOutgoingArgsTailCallable function). The actual tail call emission will call determineAssignments() again.

**Pull request merged**